### PR TITLE
feat: No disk space h5

### PR DESCRIPTION
**Pull request by VRGhost**
## Overview

This PR adds in-memory loading of HDF files.

* [vessel-state dev PR](https://github.com/VorTECHsa/vessel-state/pull/810)

## Key Changes

* passing `no_disk_space: True` for HDF file bindings will have the expected effect of no data being stored in a temporary file.

## Impact

* In-memory HDF file handling.

## Checklist

The following won't apply in all cases - mark `N/A` next to point if not needed.

- [x] Appropriate unit tests added/updated
- [x] Module, function, class docstrings updated
- [x] Comments added to PR where necessary
- [x] Interface documentation updated
- [ ] Semantic commits used for this PR, so that a CHANGELOG can be generated from them (don't delete your local branch! when tagging a new release from master)

## Performance

* dev DAG using the feature [-> link](https://airflow.kube.develop.vortexa.com/task?dag_id=vessel_state_etl_incr_no_ds&task_id=process_outcomes__kube_pod&execution_date=2022-12-15T09%3A23%3A18.429593%2B00%3A00) (duration - 2690.986829 seconds)
* unmodified DAG instance [-> link](https://airflow.kube.prod.vortexa.com/task?task_id=process_outcomes__kube_pod&dag_id=vessel_state_etl_incr&execution_date=2022-12-14T04%3A20%3A00%2B00%3A00) (duration - 2731.726569 seconds)

The modified task takes marginally shorter time to complete (-40 seconds)

## Release-Steps

- [ ] Merge PR
- [ ] Release new tag

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -556,26 +556,147 @@ def expected_messages():
...
+                "B",
+                "B",
+                "pass_through",
+                10.01234,
+                None,
+                pd.Timestamp("2021-03-30"),
+                100.01234,
+                6,
+                6,
+            ],
```
Comment (Christos-Hadjinikolis) [inline tests/conftest.py:582]: Please oh please dont reformat these!!!

Comment (VRGhost) [reply tests/conftest.py:582]: Sorry, this was an accident.

File: dynamicio/mixins/with_s3.py [hunk 2]
```diff
@@ -19,6 +25,57 @@
...
+
+    def close(self, *_args, **_kwargs):
+        pass
+
+    @property
+    def is_open(self):
+        return self._handle is not None
+
+
+class HDFIo:
```
Comment (Christos-Hadjinikolis) [inline dynamicio/mixins/with_s3.py:49]: Change name please :P

File: dynamicio/mixins/with_s3.py [hunk 3]
```diff
@@ -114,8 +183,23 @@ def _read_from_s3_path_prefix(self) -> pd.DataFrame:
...
+                    full_path_prefix,
+                    file_ext=".h5",
+                    max_memory_use=1024**3,  # 1 gib
+                ):
+                    dfs.append(HDFIo().load(fobj))
+                df = pd.concat(dfs, ignore_index=True)
+                columns = [column for column in df.columns.to_list() if column in self.schema.keys()]
+                no_disk_space_rv = df[columns]
+            if no_disk_space_rv is not None:
+                return no_disk_space_rv
```
Comment (Christos-Hadjinikolis) [inline dynamicio/mixins/with_s3.py:202]: Why not rewrite as: 
```python
        if self.options.pop("no_disk_space", False):
            if file_type == "parquet":
                return self._read_parquet_file(full_path_prefix, self.schema, **self.options)
            elif file_type == "hdf":
                dfs = []
                for fobj in self._iter_s3_files(full_path_prefix, file_ext=".h5", max_memory_use=1024**3):
                    dfs.append(HDFIo().load(fobj))
                df = pd.concat(dfs, ignore_index=True)
                columns = [column for column in df.columns.to_list() if column in self.schema.keys()]
                return df[columns]
            
```

Comment (Christos-Hadjinikolis) [reply dynamicio/mixins/with_s3.py:202]: I would also urge you to encapsulate this: 
```python
                dfs = []
                for fobj in self._iter_s3_files(full_path_prefix, file_ext=".h5", max_memory_use=1024**3):
                    dfs.append(HDFIo().load(fobj))
                df = pd.concat(dfs, ignore_index=True)
                columns = [column for column in df.columns.to_list() if column in self.schema.keys()]
                return df[columns]
``` 
into a dedicated method. But I see your point on doing this later for csv and json files too!

**Review by Christos-Hadjinikolis — CHANGES_REQUESTED** [review]
LGTM me overall, just minor changes, please see my notes.

**Review by Christos-Hadjinikolis — APPROVED** [review]

**Review by Christos-Hadjinikolis — APPROVED** [review]

**Pull request merged**